src/java.net.http/share/classes/jdk/internal/net/http/Stream.java

```diff
@@ -554,6 +553,10 @@ void incoming_reset(ResetFrame frame) {
         } else {
             Flow.Subscriber<?> subscriber =
                     responseSubscriber == null ? pendingResponseSubscriber : responseSubscriber;
+            if (frame.getErrorCode() == ResetFrame.NO_ERROR) {
```

Comment:
We should stop sending regardless of the error code provided. 
https://www.rfc-editor.org/rfc/rfc9113#name-rst_stream does not mention any special handling for specific error codes.

Reply:
But only in case of NO_ERROR should we complete the requestBodyCF successfully.
In other cases, the reception of the reset frame should be considered as an error. 
There is also a subtle issue as to *when* the reset frame should be processed, and if it comes after we have started processing DataFrame we should not be processing it before we have processed the DataFrame that came before. 

@c-cleary could you please improve the comment here to explain what we're doing, and quote the relevant part of the spec (the paragraph that explains that a server may send a RESET frame with NO_ERROR *after* having sent the complete response if it doesn't need the request body to handle the response)?

Reply:
ah, I didn't check the surrounding code.
Unless I'm missing something, if we receive a data frame followed by a reset with error code other than NO_ERROR, we will still run into this bug: incoming_reset will only enqueue the reset frame and do nothing else.
Also, if we receive the reset frame as the first frame on the stream, shouldn't we complete exceptionally, like we do today?

Reply:
> Unless I'm missing something, if we receive a data frame followed by a reset with error code other than NO_ERROR, we will still run into this bug: incoming_reset will only enqueue the reset frame and do nothing else.
Good point. Maybe we should check whether requestBodyCF.isDone(), and if not, complete the requestBodyCF exceptionally before falling through to the rest of the code (checking isDone() is just an optimization to avoid creating the IOException if it's going to be ignored).

Reply:
> Also, if we receive the reset frame as the first frame on the stream, shouldn't we complete exceptionally, like we do today?

I don't think that has changed. Even if we complete the requestBodyCF with null (which means stop sending / request body successfully sent) then the reading of the response body will kick in, and we will process the reset frame inserted in the inputQ at that point.

Reply:
> ah, I didn't check the surrounding code. Unless I'm missing something, if we receive a data frame followed by a reset with error code other than NO_ERROR, we will still run into this bug: incoming_reset will only enqueue the reset frame and do nothing else. Also, if we receive the reset frame as the first frame on the stream, shouldn't we complete exceptionally, like we do today?

You are correct in your observation here, this does indeed cause the bug to resurface as you said it would. I agree with Daniel's assessment of what should be done instead WRT checking if `requestBodyCF.isDone()` and it does fix the issue.

After testing and adjusting of the comment as mentioned, I'll publish that change if all is well.